pkg/controllers/hub/trafficmanagerprofile/controller.go

```diff
@@ -0,0 +1,196 @@
+
+// Reconcile triggers a single reconcile round.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	name := req.NamespacedName
+	profileKRef := klog.KRef(name.Namespace, name.Name)
+
+	startTime := time.Now()
+	klog.V(2).InfoS("Reconciliation starts", "trafficManagerProfile", profileKRef)
+	defer func() {
+		latency := time.Since(startTime).Milliseconds()
+		klog.V(2).InfoS("Reconciliation ends", "trafficManagerProfile", profileKRef, "latency", latency)
+	}()
+
+	profile := &fleetnetv1alpha1.TrafficManagerProfile{}
+	if err := r.Client.Get(ctx, name, profile); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.V(4).InfoS("Ignoring NotFound trafficManagerProfile", "trafficManagerProfile", profileKRef)
+			return ctrl.Result{}, nil
+		}
+		klog.ErrorS(err, "Failed to get trafficManagerProfile", "trafficManagerProfile", profileKRef)
+		return ctrl.Result{}, controller.NewAPIServerError(true, err)
+	}
+
+	if !profile.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.handleDelete(ctx, profile)
+	}
+
+	// register finalizer
+	if !controllerutil.ContainsFinalizer(profile, objectmeta.TrafficManagerProfileFinalizer) {
+		controllerutil.AddFinalizer(profile, objectmeta.TrafficManagerProfileFinalizer)
+		if err := r.Update(ctx, profile); err != nil {
+			klog.ErrorS(err, "Failed to add finalizer to trafficManagerProfile", "trafficManagerProfile", profileKRef)
+			return ctrl.Result{}, controller.NewUpdateIgnoreConflictError(err)
+		}
+	}
+
+	return r.handleUpdate(ctx, profile)
+}
+
+func (r *Reconciler) handleDelete(ctx context.Context, profile *fleetnetv1alpha1.TrafficManagerProfile) (ctrl.Result, error) {
+	profileKObj := klog.KObj(profile)
+	// The profile is being deleted
+	if !controllerutil.ContainsFinalizer(profile, objectmeta.TrafficManagerProfileFinalizer) {
+		klog.V(4).InfoS("TrafficManagerProfile is being deleted", "trafficManagerProfile", profileKObj)
+		return ctrl.Result{}, nil
+	}
+
+	klog.V(2).InfoS("Deleting Azure Traffic Manager profile", "trafficManagerProfile", profileKObj)
+	if _, err := r.ProfilesClient.Delete(ctx, r.ResourceGroupName, profile.Name, nil); err != nil {
+		if !azureerrors.IsNotFound(err) {
+			klog.ErrorS(err, "Failed to delete Azure Traffic Manager profile", "trafficManagerProfile", profileKObj)
+			return ctrl.Result{}, err
+		}
+	}
+
+	controllerutil.RemoveFinalizer(profile, objectmeta.TrafficManagerProfileFinalizer)
+	if err := r.Client.Update(ctx, profile); err != nil {
+		klog.ErrorS(err, "Failed to remove trafficManagerProfile finalizer", "trafficManagerProfile", profileKObj)
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) handleUpdate(ctx context.Context, profile *fleetnetv1alpha1.TrafficManagerProfile) (ctrl.Result, error) {
+	profileKObj := klog.KObj(profile)
+	res, updateErr := r.ProfilesClient.CreateOrUpdate(ctx, r.ResourceGroupName, profile.Name, generateAzureTrafficManagerProfile(profile), nil)
```

Comment:
We should first get if there's an existing one, then check if any field is changing and only call CreateOrUpdate when needed.